### ISSUE TITLE
is ConcurrentQueueSegment holding on to objects after being dequeued

### ISSUE THREAD

**Issue by PeterPann23** [dotnet/corefx#40571]
# is ConcurrentQueueSegment holding on to objects after being dequeued?
when looking at the memory profile of my application I noticed that I have quite a few items in the ConcurentQueue that schould nolonger be there after having dequeued them.
# General

I am using dot net core 3.0.100-preview8-013656 and get the following picture when comparing snapshots against base line

![Capture](https://user-images.githubusercontent.com/44400822/63655118-24e35700-c784-11e9-89ce-a9bc70200abd.PNG)

I'm not understanding what's going on here

**Comment by scalablecory** [comment]
If you're using `TryPeek` it will hold on to references. Otherwise, if you have some code that reproduces the issue, please provide.

**Comment by stephentoub** [comment]
> If you're using TryPeek prior to dequeueing

Or enumerating.

At that point the ones will remain referenced by the segment until the segment itself has no more items to be dequeued and the segment itself will be dropped.

**Comment by PeterPann23** [comment]
@scalablecory Yes I do TryPeek, what's the work around of how to then get rid of the queue's reference. If I can't solve it I will have to stop using this class

**Comment by PeterPann23** [comment]
Yes, I need to see if the order processing has ended by the variouse threads that are processing this order(s), hence I peek and see if it's done by all processes and then remove all subscribers. When done we store the order as all dependent data is collected. As per your comment of the item being droped, I'm not seeing this happen, not seeing any references that would block it... What I did to solve it is deque, test and enqueue it again "if not done" rather than waiting on it, I loose my order but I can solve that another way. 

I looked again at the documentation, no mention of this buffering of the peeked items.  I also do not think it's releasing the instances as when peeking as dequeuing is garbage collecting the instances.

**Comment by PeterPann23** [comment]
that's file, perhaps update the documentation and close the ticket
